File: python/fuzz_emoji.py
from typing import List, Mapping, Tuple

import emoji
import requests


class FuzzEmoji:
    def __init__(self):
        self.emoji_dict = {}
        emoji_list = {name: data for name, data in emoji.EMOJI_DATA.items() if 'en' in data}
        for emoji_char, data in emoji_list.items():
            name = data['en'].strip(':')
            self.emoji_dict[name.lower()] = emoji_char

    @staticmethod
    def get_synonyms(word):
        response = requests.get(f"https://api.datamuse.com/words?rel_syn={word}")
        if response.status_code == 200:
            synonyms = [word_data['word'] for word_data in response.json()]
            return synonyms

        raise RuntimeError(response.content)

    def get_emoji(self, description) -> Tuple[str, str]:
        description = description.lower()
        # direct match
        if description in self.emoji_dict:
            return description, self.emoji_dict[description]

        # Subset match
        for name in self.emoji_dict:
            if description in name:
                return name, self.emoji_dict[name]

        synonyms = self.get_synonyms(description)
        # Synonym match
        for syn in synonyms:
            if syn in self.emoji_dict:
                return syn, self.emoji_dict[syn]

        # Subset match
        for name in self.emoji_dict:
            for syn in synonyms:
                if syn in name:
                    return syn, self.emoji_dict[name]
        return '', ''

    def get_emojis(self, descriptions: List[str]) -> Mapping[str, str]:
        return {d: str(self.get_emoji(d)) for d in descriptions}


def main():
    descriptions = ["flame", "confused", "green", "Israel"]
    fuzzer = FuzzEmoji()
    result = fuzzer.get_emojis(descriptions)
    for k, v in result.items():
        print(k, v)


if __name__ == '__main__':
    main()

File: rust/src/fuzz_emoji.rs
extern crate reqwest;
extern crate serde;
extern crate serde_json;
extern crate emojis;

use emojis::iter;
use std::collections::HashMap;
use std::error::Error;
use serde::Deserialize;

#[derive(Debug)]
pub(crate) struct FuzzEmoji {
    emoji_dict: HashMap<String, String>,
}

impl FuzzEmoji {
    pub(crate) fn new() -> FuzzEmoji {
        let mut emoji_dict = HashMap::new();
        for emoji in iter() {
            let name = emoji.shortcodes().next().unwrap_or_default().to_string();
            let glyph = emoji.as_str().to_string();
            emoji_dict.insert(name, glyph);
        }
        FuzzEmoji { emoji_dict }
    }

    fn get_synonyms(&self, word: &str) -> Result<Vec<String>, Box<dyn Error>> {
        let url = format!("https://api.datamuse.com/words?rel_syn={}", word);
        let resp = reqwest::blocking::get(&url)?;

        if resp.status().is_success() {
            let body = resp.text()?;
            let words: Vec<Word> = serde_json::from_str(&body)?;
            let synonyms = words.into_iter().map(|w| w.word).collect();
            Ok(synonyms)
        } else {
            Err(format!("failed to fetch synonyms: {}", resp.status()).into())
        }
    }

    fn get_emoji(&self, description: &str) -> (String, String) {
        let description = description.to_lowercase();

        if let Some(emoji_char) = self.emoji_dict.get(&description) {
            return (description.clone(), emoji_char.clone());
        }

        for (name, emoji_char) in &self.emoji_dict {
            if name.contains(&description) {
                return (name.clone(), emoji_char.clone());
            }
        }

        let synonyms = match self.get_synonyms(&description) {
            Ok(synonyms) => synonyms,
            Err(_) => return ("".to_string(), "".to_string()),
        };

        for syn in &synonyms {
            if let Some(emoji_char) = self.emoji_dict.get(syn) {
                return (syn.clone(), emoji_char.clone());
            }
        }

        for (name, emoji_char) in &self.emoji_dict {
            for syn in &synonyms {
                if name.contains(syn) {
                    return (syn.clone(), emoji_char.clone());
                }
            }
        }

        ("".to_string(), "".to_string())
    }

    pub(crate) fn get_emojis(&self, descriptions: Vec<&str>) -> HashMap<String, String> {
        let mut result = HashMap::new();
        for d in descriptions {
            let (name, emoji_char) = self.get_emoji(d);
            result.insert(d.to_string(), format!("{}, {}", name, emoji_char));
        }
        result
    }
}

#[derive(Deserialize)]
struct Word {
    word: String,
}